File: app/src/main/java/com/jqyd/yuerduo/net/GsonHttpResponseHandler.java
package com.jqyd.yuerduo.net;

import com.loopj.android.http.BaseJsonHttpResponseHandler;
import com.orhanobut.logger.Logger;

import java.lang.reflect.ParameterizedType;

import cz.msebera.android.httpclient.Header;

/**
 * gson解析处理器
 * Created by zhangfan on 2016/1/25.
 */
public abstract class GsonHttpResponseHandler<T> extends BaseJsonHttpResponseHandler<ResultHolder<T>> {

    @Override
    public void onSuccess(int statusCode, Header[] headers, String rawJsonResponse, ResultHolder<T> response) {
//        Logger.i(""+rawJsonResponse);
//        Logger.json(rawJsonResponse);
    }

    @Override
    public void onFailure(int statusCode, Header[] headers, Throwable throwable, String rawJsonData, ResultHolder<T> errorResponse) {
//        Logger.i(""+rawJsonData);
//        Logger.e(throwable.getMessage());
    }

    @Override
    protected final ResultHolder<T> parseResponse(String rawJsonData, boolean isFailure) throws Throwable {
        if (isFailure) return null;
        Class<T> entityClass = (Class<T>) ((ParameterizedType) this.getClass().getGenericSuperclass()).getActualTypeArguments()[0];
        return ResultHolder.fromJson(rawJsonData, entityClass);
    }

    @Override
    public void onRetry(int retryNo) {
        super.onRetry(retryNo);
        Logger.i(String.format("Request retry no. %d", retryNo));
    }
}

File: app/src/main/java/com/jqyd/yuerduo/activity/common/GsonHttpResponseHandler.kt
package com.jqyd.yuerduo.activity.common

import com.jqyd.yuerduo.net.ResultHolder
import com.loopj.android.http.BaseJsonHttpResponseHandler
import com.orhanobut.logger.Logger
import cz.msebera.android.httpclient.Header

/**
 * gson解析处理器
 * Created by zhangfan on 2016/1/25.
 */
abstract class GsonHttpResponseHandler<T> : BaseJsonHttpResponseHandler<ResultHolder<T>>() {

    override fun onSuccess(statusCode: Int, headers: Array<out Header>?, rawJsonResponse: String?, response: ResultHolder<T>?) {

    }

    override fun onFailure(statusCode: Int, headers: Array<out Header>?, throwable: Throwable?, rawJsonData: String?, errorResponse: ResultHolder<T>?) {

    }

    abstract fun getCls():Class<*>

    @Throws(Throwable::class)
    override fun parseResponse(rawJsonData: String, isFailure: Boolean): ResultHolder<T>? {
        if (isFailure) return null
        return ResultHolder.fromJson<T>(rawJsonData, getCls())
    }

    override fun onRetry(retryNo: Int) {
        super.onRetry(retryNo)
        Logger.i(String.format("Request retry no. %d", retryNo))
    }
}
